pkg/nsx/services/ipaddressallocation/ipaddressallocation.go

```diff
@@ -0,0 +1,197 @@
+		return false, err
+	}
+	log.V(1).Info("existing ipaddressallocation", "ipaddressallocation", existingIPAddressAllocation)
+	_ = common.CompareResource(IpAddressAllocationToComparable(existingIPAddressAllocation),
+		IpAddressAllocationToComparable(nsxIPAddressAllocation))
+
+	//if !ipAddressAllocationUpdated{
+	//	return false, nil
+	//}
+
+	if err := service.Apply(nsxIPAddressAllocation); err != nil {
+		return false, err
+	}
+
+	createdIPAddressAllocation, err := service.indexedIPAddressAllocation(obj.UID)
+	if err != nil {
+		log.Error(err, "failed to get ipaddressallocation", "UID", obj.UID)
+		return false, err
+	}
+	cidr := createdIPAddressAllocation.AllocationIps
+	if cidr == nil {
+		return false, fmt.Errorf("ipaddressallocation %s didn't realize available cidr", obj.UID)
+	}
+	obj.Status.CIDR = *cidr
+	return true, nil
+}
+
+func (service *IPAddressAllocationService) Apply(nsxIPAddressAllocation *model.VpcIpAddressAllocation) error {
+	ns := service.GetIPAddressAllocationNamespace(nsxIPAddressAllocation)
+	var err error
+	VPCInfo := service.VPCService.ListVPCInfo(ns)
+	if len(VPCInfo) == 0 {
+		err = util.NoEffectiveOption{Desc: "no valid org and project for ipaddressallocation"}
+		return err
+	}
+	err = service.NSXClient.IPAddressAllocationClient.Patch(VPCInfo[0].OrgID, VPCInfo[0].ProjectID, VPCInfo[0].ID, *nsxIPAddressAllocation.Id, *nsxIPAddressAllocation)
+	err = util.NSXApiError(err)
+	if err != nil {
+		// not return err, try to get it from nsx, in case if cidr not realized at the first time
+		// support realize multiple times
+		log.Error(err, "patch failed, try to get it from nsx")
+	}
+	// get back from nsx, it contains path which is used to parse vpc info when deleting
+	nsxIPAddressAllocationNew, err := service.NSXClient.IPAddressAllocationClient.Get(VPCInfo[0].OrgID, VPCInfo[0].ProjectID, VPCInfo[0].ID, *nsxIPAddressAllocation.Id)
+	err = util.NSXApiError(err)
+	if err != nil {
+		return err
+	}
+	if nsxIPAddressAllocation.AllocationIps == nil {
+		err := fmt.Errorf("cidr not realized yet")
+		return err
+	}
+	err = service.ipAddressAllocationStore.Apply(&nsxIPAddressAllocationNew)
+	if err != nil {
+		return err
+	}
+	log.V(1).Info("successfully created or updated ipaddressallocation", "nsxIPAddressAllocation", nsxIPAddressAllocation)
+	return nil
+}
+
+func (service *IPAddressAllocationService) DeleteIPAddressAllocation(obj interface{}) error {
+	var err error
+	var nsxIPAddressAllocation *model.VpcIpAddressAllocation
+	switch o := obj.(type) {
+	case *v1alpha1.IPAddressAllocation:
+		nsxIPAddressAllocation, err = service.indexedIPAddressAllocation(o.UID)
+		if err != nil {
+			log.Error(err, "failed to get ipaddressallocation", "IPAddressAllocation", o)
+			return err
+		}
+	case types.UID:
+		nsxIPAddressAllocation, err = service.indexedIPAddressAllocation(o)
+		if err != nil {
+			log.Error(err, "failed to get ipaddressallocation by UID", "UID", o)
+			return err
+		}
+	}
+	if nsxIPAddressAllocation == nil {
```

Comment:
Add a log to indicate the object not found before returning nil?

Reply:
ok